### PR TITLE
feat(ui): improve types

### PR DESCRIPTION
**Pull request by Shinigami92**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Documentation
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch (or `v[X]` branch)
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] It's been tested on a Cordova (iOS, Android) app
- [ ] It's been tested on a Electron app
- [ ] Any necessary documentation has been added or updated [in the docs](https://github.com/quasarframework/quasar/tree/dev/docs) (for faster update click on "Suggest an edit on GitHub" at bottom of page) or explained in the PR's description.

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

While starting to build an electron-quasar app, I experienced some lazy defined types and I want to improve the auto-completion in VSCode for TypeScript users like me.

https://github.com/Shinigami92/watchlist/blob/main/src/quasar.d.ts

So @yusufkandemir suggested that I can update these types by changing the json schema files.
This may be just the start from my side to make some contributions to Quasar :raised_hands: 
Please feel free to push me in the right direction if I do anything wrong :slightly_smiling_face:

### CODE REVIEW

File: ui/src/components/option-group/QOptionGroup.json [hunk 1]
```diff
@@ -15,6 +15,31 @@
     "options": {
       "type": "Array",
       "desc": "Array of objects with value and label props. The binary components will be created according to this array; Props from QToggle, QCheckbox or QRadio can also be added as key/value pairs to control the components singularly",
+      "definition": {
+        "label": {
+          "type": "String",
+          "desc": "Label",
```
Comment (Shinigami92) [inline ui/src/components/option-group/QOptionGroup.json]: Please help me to fine a better descriptive text

File: ui/src/components/option-group/QOptionGroup.json [hunk 2]
```diff
@@ -15,6 +15,31 @@
...
+          "required": true,
+          "examples": [ "Option 1", "Option 2", "Option 3" ]
+        },
+        "value": {
+          "type": "Any",
+          "desc": "Value",
```
Comment (Shinigami92) [inline ui/src/components/option-group/QOptionGroup.json]: Please help me to fine a better descriptive text

File: ui/src/components/option-group/QOptionGroup.json [hunk 3]
```diff
@@ -15,6 +15,31 @@
...
+          "required": true,
+          "examples": [ "op1", "op2", "op3" ]
+        },
+        "disable": {
+          "type": "Boolean",
+          "desc": "Disable",
```
Comment (Shinigami92) [inline ui/src/components/option-group/QOptionGroup.json]: Please help me to fine a better descriptive text

**Comment by Shinigami92** [comment]
@IlCallo You wrote me on Discord (https://discord.com/channels/415874313728688138/807654245640962059/901927932237348886)

What I would like is to create types that are more enhanced.

So instead of this (_JSDoc comments omitted_):

```ts
export interface QOptionGroupProps {
  size?: string | undefined;
  modelValue: any;
  options?:
    | {
        label: string;
        value: any;
        disable?: boolean;
        [index: string]: any;
      }[]
    | undefined;
  name?: string | undefined;
  type?: "radio" | "checkbox" | "toggle" | undefined;
  color?: string | undefined;
  keepColor?: boolean | undefined;
  dark?: boolean | undefined;
  dense?: boolean | undefined;
  leftLabel?: boolean | undefined;
  inline?: boolean | undefined;
  disable?: boolean | undefined;
  "onUpdate:modelValue"?: (value: any) => void;
}
```

I would like to create types like this:

```ts
export interface QOptionGroupProps<Value = any> {
  size?: string | undefined;
  modelValue: Value;
  options?:
    | {
        label: string;
        value: Value;
        disable?: boolean;
        [index: string]: any;
      }[]
    | undefined;
  name?: string | undefined;
  type?: "radio" | "checkbox" | "toggle" | undefined;
  color?: string | undefined;
  keepColor?: boolean | undefined;
  dark?: boolean | undefined;
  dense?: boolean | undefined;
  leftLabel?: boolean | undefined;
  inline?: boolean | undefined;
  disable?: boolean | undefined;
  "onUpdate:modelValue"?: (value: Value) => void;
}
```

And this is just one example...

So what do I have to do to define such types?

File: ui/src/components/option-group/QOptionGroup.json [hunk 4]
```diff
@@ -15,6 +15,31 @@
...
+          "default": false,
+          "examples": [ true, false ]
+        },
+        "...props": {
+          "type": "Any",
+          "desc": "Any other QOptionGroup props",
```
Comment (yusufkandemir) [inline ui/src/components/option-group/QOptionGroup.json]:
```suggestion
          "desc": "Any other props from QToggle, QCheckbox, or QRadio",
```

**Comment by Shinigami92** [comment]
Another example for `QTableProps`

```ts
export interface QTableProps<
    Row extends Record<string, any> = any,
    K = Row extends Record<string, any> ? keyof Row : string,
    Field = K | ((row: Row) => any)
  > {
  // ...
  rows?: Row[] | undefined;
  rowKey?: string | ((row: Row) => string) | undefined;
  // ...
  columns?:
    | {
        // ...
        field: Field;
        // ...
        align?: "left" | "right" | "center";
        // ...
        sort?: (a: any, b: any, rowA: Row, rowB: Row) => number;
        // ...
        format?: (val: any, row: Row) => string;
        style?: string | ((row: Row) => string);
        classes?: string | ((row: Row) => string);
        // ...
      }[]
    | undefined;
  // ...
  // And many more can be enhanced here
}
```

---

I also would prefer to extract the `columns` type, so devs can directly use `QTableColumn[]`

File: ui/src/components/option-group/QOptionGroup.json (hunk 3 above)
Comment (yusufkandemir) [reply ui/src/components/option-group/QOptionGroup.json]:
```suggestion
          "desc": "If true, the option will be disabled",
```

File: ui/src/components/option-group/QOptionGroup.json (hunk 2 above)
Comment (yusufkandemir) [reply ui/src/components/option-group/QOptionGroup.json]:
```suggestion
          "desc": "Value of the option that will be used by the component model",
```

File: ui/src/components/option-group/QOptionGroup.json (hunk 1 above)
Comment (yusufkandemir) [reply ui/src/components/option-group/QOptionGroup.json]:
```suggestion
          "desc": "Label to display along the component",
```

File: ui/src/components/option-group/QOptionGroup.json [hunk 5]
```diff
@@ -15,6 +15,31 @@
...
+          "type": "Any",
+          "desc": "Value",
+          "required": true,
+          "examples": [ "op1", "op2", "op3" ]
+        },
+        "disable": {
+          "type": "Boolean",
+          "desc": "Disable",
+          "default": false,
+          "examples": [ true, false ]
```
Comment (yusufkandemir) [inline ui/src/components/option-group/QOptionGroup.json]:
```suggestion
```
Boolean type is obvious and doesn't need examples.

File: ui/src/components/option-group/QOptionGroup.json [hunk 6]
```diff
@@ -15,6 +15,31 @@
...
+        },
+        "...props": {
+          "type": "Any",
+          "desc": "Any other QOptionGroup props",
+          "__exemption": [ "examples" ]
```
Comment (yusufkandemir) [inline ui/src/components/option-group/QOptionGroup.json]: `examples` do not have any examples of props from other components(_QToggle, QCheckbox, or QRadio_). Please try to add some.

**Review by yusufkandemir — APPROVED** [review]
👏

File: ui/src/components/option-group/QOptionGroup.json (hunk 6 above)
Comment (Shinigami92) [reply ui/src/components/option-group/QOptionGroup.json]: What especially is this `"__exemption"`? I copied it over from somewhere else :sweat_smile:  
I just discovered that it is possible to add some custom props like in this example: https://quasar.dev/vue-components/option-group#example--label-slots  

So should it be this?

```suggestion
          "__exemption": [ "icon", "color" ]
```

Comment (hawkeye64) [reply ui/src/components/option-group/QOptionGroup.json]: In the associated JSON file, `__exception` means the targeted props are not required to have an example.

Comment (yusufkandemir) [reply ui/src/components/option-group/QOptionGroup.json]: Let's say you have a function definition, which has `examples`, and for a specific param of it, you think the `examples` from the function are already enough to describe it, so you can add `examples` to `__exemption` to skip the need of adding an example for that specific param.

However, in this case, `examples` of `options` are not descriptive enough, so you shouldn't skip defining an example. So, you would need to remove the `__exemption` completely and add `examples` prop with some values like `"examples": [ "whatever=\"foo-bar\"" ]` but with a better example than `whatever` 😄.

Comment (yusufkandemir) [reply ui/src/components/option-group/QOptionGroup.json]:
```suggestion
          "examples": [ "val=\"car\"", ":true-value=\"trueValue\"", "checked-icon=\"visibility\"" ]
```
I think these examples should be clear enough.

File: ui/src/components/option-group/QOptionGroup.json [hunk 7]
```diff
@@ -15,6 +15,30 @@
     "options": {
       "type": "Array",
       "desc": "Array of objects with value and label props. The binary components will be created according to this array; Props from QToggle, QCheckbox or QRadio can also be added as key/value pairs to control the components singularly",
```
Comment (yusufkandemir) [inline ui/src/components/option-group/QOptionGroup.json]:
```suggestion
      "desc": "Array of objects with value, label, and disable(optional) props. The binary components will be created according to this array; Props from QToggle, QCheckbox or QRadio can also be added as key/value pairs to control the components singularly",
```
Let's clarify the description a bit while we are at it 🙂

**Comment by Shinigami92** [comment]
I think this PR is ready to get merged.
I can still open further PRs in the future whenever I encounter something I need to enhance :slightly_smiling_face:

**Review by yusufkandemir — APPROVED** [review]
Thanks for the contribution, nicely done 👏

**Pull request merged**